### DISCUSSION TITLE
Is there any way to execute some code (saving activity details in db) after executing activity successfully in ELSA?

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#4973] [Q&A]
I wanted to save activity details in mongo db after execution of particular activity which may be in different state like executed, faulted, Suspsened, idle etc? I have impleted workflow by implementing IWorkflfow interface and activities by implementing Elsa.Services.Activity? 
Is there any way(like middleware or events or some other way) to execute some code which will be mostly used for saving activity details in db?

**Comment by glime-ai** [comment]
To save activity details in MongoDB after the execution of a particular activity in Elsa, you can create a custom activity that inherits from `Activity` and overrides the `OnExecutedAsync` method to include logic for saving the activity details to MongoDB. You can also handle other states like faulted or suspended by overriding the respective methods such as `OnFaultedAsync` or `OnSuspendedAsync`.

Here's a simplified example of what your custom activity might look like:

```csharp
public class MyCustomActivity : Activity
{
    private readonly IMongoCollection<ActivityDetails> _activityDetailsCollection;

    public MyCustomActivity(IMongoCollection<ActivityDetails> activityDetailsCollection)
    {
        _activityDetailsCollection = activityDetailsCollection;
    }

    protected override async ValueTask<IActivityExecutionResult> OnExecutedAsync(ActivityExecutionContext context)
    {
        // Logic to save activity details after execution
        var activityDetails = new ActivityDetails
        {
            ActivityId = context.ActivityId,
            WorkflowInstanceId = context.WorkflowInstance.Id,
            State = "Executed",
            // Other details you want to save
        };

        await _activityDetailsCollection.InsertOneAsync(activityDetails);

        return Done();
    }

    protected override async ValueTask<IActivityExecutionResult> OnFaultedAsync(ActivityExecutionContext context, Exception exception)
    {
        // Logic to save activity details when faulted
        var activityDetails = new ActivityDetails
        {
            ActivityId = context.ActivityId,
            WorkflowInstanceId = context.WorkflowInstance.Id,
            State = "Faulted",
            // Other details you want to save
        };

        await _activityDetailsCollection.InsertOneAsync(activityDetails);

        return Fault(exception);
    }

    // Similarly, you can override OnSuspendedAsync, etc.
}
```

In this example, `ActivityDetails` is a class that you would define to represent the details you want to store in MongoDB. You would also need to inject the `IMongoCollection<ActivityDetails>` into your custom activity, which would be set up in your application's dependency injection container.

Remember to register your custom activity with Elsa so that it can be used within your workflows. You can do this by implementing `IActivityProvider` or by registering it directly in the service collection during application startup.

Please note that this is a high-level example and you will need to adapt it to fit your specific requirements and the structure of your MongoDB collections.
